### ISSUE TITLE
Parsing error when using template type parameter with assignment

### ISSUE THREAD

**Issue by slnj** [SonarOpenCommunity/sonar-cxx#2317]
Hello!

When using a template parameter with regular type parameters with default value assignment, a parsing error is reported.
This cases are valid and generate parsing errors:
```
template < typename T,  size_t s = 15 >
void f(){}
```
```
template < values v = v::ok >   // where values is an enum
void f(){}
```
```
template<class T, T::type n = 0>  //Where T::type is an integral type
class X;
```
The type could be:
  integral type
  enumerator
  lvalue reference
  nullptr
  pointer

An a default value could be assigned

Also, the default value could be an expression like:  (sizeof(T)*2)

Regards

**Comment by guwirth** [comment]
Hello @slnj,

will have a look to it.

For your explanation:

General problem is that the cxx plugin is only parsing the grammar (syntax analysis) without knowing the semantic:

Below is a good sample: First case works because `int` is an integral type. In second case `size_t` is for the parser just a identifier and only from the position the parser can guess that it is a type.
```C++
template < typename T,  int s = 15 > void f(){} // works: int is integral type
template < typename T,  size_t s = 15 > void f(){} // fails
```
Our second problem is that we are using [SonarSource Language Recognizer (SSLR)](https://github.com/SonarSource/sslr) which implements [PEG parsing](https://en.wikipedia.org/wiki/Parsing_expression_grammar). Syntactically, PEGs also look similar to [context-free grammars](https://en.wikipedia.org/wiki/Context-free_grammar) (CFGs), but they have a different interpretation: the choice operator selects the first match in PEG, while it is ambiguous in CFG. That means we have to change the order of the rules in the grammar in some places. This is ultimately a lot of "try and error", especially in such a complex grammar as C++.

Regards,
